### DISCUSSION TITLE
Screenshot: Reading from the front buffer is unreliable

### DISCUSSION THREAD

**Discussion by gokhalen** [GLVis/glvis#257] [General]
Upon using the command in a script file

```
solution NHC_NHG_prom_onlinehyp-mesh1.000000 error-online0.000000 screenshot 1.png
```

I get 

```
MFEM Warning: Screenshot: Reading from the front buffer is unreliable.  Resulting screenshots may be incorrect.

 ... in function: int Screenshot(const char*, bool)
 ... in file: lib/aux_vis.cpp:1018

-> 1.png
```

and `1.png` is junk. And ideas on how to fix?

FWIW, I'm using it on WSL2 on Windows10 with the VcXsrv X Server.

**Comment by v-dobrev** [comment]
Thank you for reporting this issue.

This maybe similar to the issue reported here: https://github.com/GLVis/glvis/issues/254.

As a workaround, you can try adding a no-op command like `keys ~~` (which may force a redraw, I'm not sure) just before the `screenshot` command. Let us know if that works or not.

**Reply by gokhalen** [reply]
No, what you suggested doesn't seem to work.

**Reply by v-dobrev** [reply]
Okay, thanks. We'll need to find some time to debug this.

One approach that should work, but will require you to press space multiple times, is to group loading the mesh/solution into one group and then the screenshot in another:
```
{ solution NHC_NHG_prom_onlinehyp-mesh1.000000 error-online0.000000 }
{ screenshot 1.png }
```
Basically, at the end of the group, GLVis will stop and wait for the next space key input which will allow it to update the window content before taking the screenshot.

**Reply by gokhalen** [reply]
That doesn't work either, unfortunately.  I get the same error as before. 
My script is 

```
# Visualization window geometry
window 0 0 600 600

solution NHC_NHG_prom_onlinehyp-mesh1.000000 NHC_NHG_prom_onlinehyp-final0.000000

# Setup the GLVis scene. Executed after pressing the space bar.
{
   perspective off
   view 0 0
   viewcenter 0 0
   zoom 1.0
}

keys b

{ solution NHC_NHG_prom_onlinehyp-mesh1.000000 NHC_NHG_prom_onlinehyp-final1.000000 }
{ screenshot dummy1.png }

keys b

{ solution NHC_NHG_prom_onlinehyp-mesh1.000000 NHC_NHG_prom_onlinehyp-final2.000000 }
{ screenshot dummy2.png }

keys b

{ solution NHC_NHG_prom_onlinehyp-mesh1.000000 NHC_NHG_prom_onlinehyp-final3.000000 }
{ screenshot dummy3.png }
```

**Reply by v-dobrev** [reply]
I was able to create 2 screenshots from the results of running `ex9` using this script:
```
# Window position and size
window 0 0 600 600

# load initial mesh and solution
solution ex9.mesh ex9-init.gf

# this section adjusts the view, zoom, what is visible, etc.
{ keys aA }

{ screenshot ex9-01.png }
{ solution ex9.mesh ex9-final.gf }
{ screenshot ex9-02.png }
```
and pressing space 4 times after running it with `./glvis -run ex9.glvs`. I do get the MFEM warning but the screenshots are actually correct.

**Reply by gokhalen** [reply]
Your script doesn't work for me. I get the error and junk in the png files. Perhaps some weird interaction with WSL + VcXsrv. Anyway, I'm currently generating screenshots using command line options.

**Reply by v-dobrev** [reply]
If you want, you can also try the native Windows app from https://glvis.org ("Windows" link below the download button). I'm not sure if that will work better though.

**Reply by najlkin** [reply]
A bit late response 😅 , but it works for me to under Windows with WSL to use `keys S` instead of `screenshot`. The implementation is a bit different internally and the former works more reliably. Anyway, we would like to fix this in the near future in connection with some new features 🚀
